app/src/main/java/fr/free/nrw/commons/contributions/ContributionController.java

```diff
@@ -46,11 +56,105 @@ public void initiateCameraPick(Activity activity) {
 
         PermissionUtils.checkPermissionsAndPerformAction(activity,
                 Manifest.permission.WRITE_EXTERNAL_STORAGE,
-                () -> initiateCameraUpload(activity),
+                () -> askUserToAttachLocation(activity),
                 R.string.storage_permission_title,
                 R.string.write_storage_permission_rationale);
     }
 
+    /**
+     * Suggest user to attach location information with pictures.
+     * If the user selects "Yes", then:
+     *
+     * Location is taken from the EXIF if the default camera application
+     * does not redact location tags.
+     *
+     * Otherwise, if the EXIF metadata does not have location information, then location captured by the app is used
+     *
+     * If the user selects "No", then the location tag in EXIF is also removed.
```

Comment:
Am I understanding correctly that this removes location from EXIF?

If the picture has EXIF location, the app should just use it silently (no need to show this dialog).

_Sorry if your implementation was influenced by my recent creation of https://github.com/commons-app/apps-android-commons/issues/5247, you really are not at all required to implement that one._

Reply:
Yes, this removes location from EXIF. My implementation was _not_ influenced by issue #5247 at all. Pull requests are supposed to be specific, I would have raised another one for that issue. I coincidentally implemented it while implementing [this](https://github.com/commons-app/apps-android-commons/issues/4757#issuecomment-1021906251) dialog. I'm sorry, I should have discussed about this dialog because metadata is crucial and I shouldn't have simply removed the tags without asking😔

Elaborating why I felt this was the intended behaviour:

> If the picture has EXIF location, the app should just use it silently

We cannot predict if some camera application redacts EXIF location tags unless we have the picture ready with us in the Upload wizard. And if we need to record location before image capture, we need to have access to user's location. For that, a dialog is necessary as we need to explain to the users why they need to provide the access even while using camera functionality.

We can make it appear once with an checkbox for _Do not ask again_. I doubt if we can remove this dialog altogether.

The dialog has a "No, do not attach location" option. If the users tap on it, I think they would believe their location will not be shared along with the picture. If we still take it from the EXIF, then this would be quite misleading as location will still be attached to the picture. 

If we drop this idea of taking location before capture to detect if the image has EXIF location and record the current location only after we have the image with us, then again, we're showing a "No location found" dialog where users can simply press the target icon to point to the current location (I know this is not working correctly right now but I can fix this). I do not think we need any further changes if we're recording location only after the image is clicked :)

I'm sorry again for these assumptions and the inconvenience these may have caused.

Reply:
I had not thought of this, thanks a lot for the detailed and very understandable explanation, that helps a lot! :-)

> "_a dialog is necessary as we need to explain to the users why they need to provide the access_"

How about wording this dialog like this?

> Allow the app to fetch location in case the camera does not record it. Some device cameras do not record location. In such cases, letting the app fetch and attach location to it makes your contribution more useful.

I believe that wording it like this allows us to continue leaving EXIF untouched in cases where the EXIF actually has location.

Reply:
Thank you for sharing the dialog text. I have modified the code accordingly. An additional check that I've added is for the unchecked location option in the Manage EXIF Tags setting, as mentioned [here](https://github.com/commons-app/apps-android-commons/issues/4757#issuecomment-1019455442).